infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogDataStore.java

```diff
@@ -1,119 +1,276 @@
-            STARTING_ADDRESS_PREFIX, STARTING_ADDRESS_KEY, Long.class
+    private static final KvRecord<Long> HEAD_SEGMENT_RECORD = new KvRecord<>(
+            HEAD_SEGMENT_PREFIX, HEAD_SEGMENT_KEY, Long.class
+    );
+
+    private static final KvRecord<Long> COMPACTION_MARK_RECORD = new KvRecord<>(
+            COMPACTION_MARK_PREFIX, COMPACTION_MARK_KEY, Long.class
     );
 
+    private static final KvRecord<Long> COMMITTED_TAIL_RECORD = new KvRecord<>(
+            COMMITTED_TAIL_PREFIX, COMMITTED_TAIL_KEY, Long.class);
+
+    private static final KvRecord<Boolean> REQUIRE_STATE_TRANSFER_RECORD = new KvRecord<>(
+            REQUIRE_STATE_TRANSFER_PREFIX, REQUIRE_STATE_TRANSFER_KEY, Boolean.class);
+
     private static final long ZERO_ADDRESS = 0L;
 
     @NonNull
     private final KvDataStore dataStore;
 
     /**
-     * Cached starting address
+     * Cached starting address.
+     */
+    private final AtomicReference<Long> headSegment;
+
+    /**
+     * Cached tail segment.
+     */
+    private final AtomicReference<Long> tailSegment;
+
+    /**
+     * Cached global compaction mark.
+     * Any snapshot read before this may result in in-complete history.
+     */
+    private final AtomicReference<Long> globalCompactionMark;
+
+    /**
+     * Cached committed log tail, up to which the log is consolidated.
      */
-    private final AtomicLong startingAddress = new AtomicLong(Address.NON_ADDRESS);
+    private final AtomicReference<Long> committedTail;
+
     /**
-     * Cached tail segment
+     * Cached flag to indicate if required state transfer (state is lost)
+     * possibly due to reset operation.
      */
-    private final AtomicLong tailSegment = new AtomicLong(Address.NON_ADDRESS);
+    private final AtomicReference<Boolean> requireStateTransfer;
+
+    public StreamLogDataStore(KvDataStore dataStore) {
+        this.dataStore = dataStore;
+        this.headSegment =
+                new AtomicReference<>(dataStore.get(HEAD_SEGMENT_RECORD, Address.MAX));
+        this.tailSegment =
+                new AtomicReference<>(dataStore.get(TAIL_SEGMENT_RECORD, ZERO_ADDRESS));
+        this.globalCompactionMark =
+                new AtomicReference<>(dataStore.get(COMPACTION_MARK_RECORD, Address.NON_ADDRESS));
+        this.committedTail =
+                new AtomicReference<>(dataStore.get(COMMITTED_TAIL_RECORD, Address.NON_ADDRESS));
+        this.requireStateTransfer =
+                new AtomicReference<>(dataStore.get(REQUIRE_STATE_TRANSFER_RECORD, false));
+    }
 
     /**
-     * Return current cached tail segment or get the segment from the data store if not initialized
+     * Return the current tail segment.
      *
      * @return tail segment
      */
-    public long getTailSegment() {
-        if (tailSegment.get() == Address.NON_ADDRESS) {
-            tailSegment.set(dataStore.get(TAIL_SEGMENT_RECORD, ZERO_ADDRESS));
-        }
-
+    long getTailSegment() {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651450)